Java sprint 19 v1

1EVILGUN1 opened the pull request:
Работа в Collector:
Созданы классы наследники и схемы генерации классов SensorEvent и HubEvent

1EVILGUN1 committed:
Java sprint 19 v1

Работа в Collector:
Созданы классы наследники и схемы генерации классов SensorEvent и HubEvent

1EVILGUN1 committed:
update

Добавление config и отправка сообщений из PostMapping

1EVILGUN1 committed:
feat: create service,serializers

Переделан классы hub,sensor,controller
Добавлены классы serializers, service, exception

VadimZharkov commented on telemetry/collector/src/main/java/ru/yandex/practicum/controller/CollectorController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @PostMapping("/sensors")
+    public void processingSensors(@Valid @RequestBody SensorEvent event) {
+        switch (event.getClass().getSimpleName()) {
```
У событий есть метод getType(), его использование будет работать быстрей, чем getClass().getSimpleName() и далее сравнение со строкой

VadimZharkov commented on telemetry/collector/src/main/java/ru/yandex/practicum/service/CollectorService.java:
```diff
@@ -0,0 +1,204 @@
+@RequiredArgsConstructor
+public class CollectorService {
+
+    public void processingSensors(ClimateSensorEvent sensorEvent) {
```
Все методы имеют однотипную логику, отличаются только сериализацией. Поэтому общую логику лучше вынести в базовый параметризированный класс с абстрактным методом для сериализации

VadimZharkov commented on telemetry/collector/src/main/java/ru/yandex/practicum/service/CollectorService.java:
```diff
@@ -0,0 +1,204 @@
+        config.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ClimateSensorAvroSerializer.class);
+
+        Producer<String, ClimateSensorAvro> producer = new KafkaProducer<>(config);
+        String topic = "telemetry.sensors.v1";
```
Название топиков лучше вынести в application.yaml, т.к. потенциально они могут быть другими в разных средах развертывания

VadimZharkov commented on telemetry/collector/src/main/java/ru/yandex/practicum/service/CollectorService.java:
```diff
@@ -0,0 +1,204 @@
+        Properties config = getPropertiesSensor();
+        config.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ClimateSensorAvroSerializer.class);
+
+        Producer<String, ClimateSensorAvro> producer = new KafkaProducer<>(config);
```
Лучше сделать один продьюсер на приложение

VadimZharkov commented on telemetry/collector/src/main/java/ru/yandex/practicum/service/CollectorService.java:
```diff
@@ -0,0 +1,204 @@
+
+        ProducerRecord<String, ClimateSensorAvro> record = new ProducerRecord<>(topic, event);
+
+        producer.send(record);
```
Кроме данных желательно еще писать ключ события (за ключ взять hubId) и временную метку (взять из события).
Это позволит записи сохраняться в партицию в зависимости от id хаба, а это в свою очередь означает, что записи относящиеся к одному хабу можно будет читать упорядоченно т.к. кафка гарантирует очередность сообщений только в рамках партиции. Таймстемп записи позволит кафка упорядочивать записи по времени возникновения события, а не времени когда брокер кафки получил сообщение.

VadimZharkov commented on telemetry/collector/src/main/java/ru/yandex/practicum/serialize/sensor/ClimateSensorAvroSerializer.java:
```diff
@@ -0,0 +1,49 @@
+
+
+    @Override
+    public byte[] serialize(String topic, ClimateSensorAvro event) {
```
Во всех классах сериализаторов идентичные методы, лучше сделать один класс `GeneralAvroSerializer implements Serializer<SpecificRecordBase> `и у него реализовать public byte[] serialize(String topic, SpecificRecordBase data)

VadimZharkov commented on telemetry/collector/src/main/java/ru/yandex/practicum/service/CollectorService.java:
```diff
@@ -0,0 +1,204 @@
+        String topic = "telemetry.sensors.v1";
+
+        ProducerRecord<String, ClimateSensorAvro> record = new ProducerRecord<>(topic, event);
+
```
Хорошо бы добавить логирование
log.trace("Сохраняю событие {} связанное с хабом {} в топик {}", ...);

VadimZharkov reviewed (changes requested):
В целом хорошее начало, нужно исправить замечания

1EVILGUN1 committed:
feat delete repeating code and logger in service

Удалены классы сериализации avro. Создан обобщающие классы Process, GeneralAvroSerializer. Добавлены логи в класс service
